File: shakey-server/src/main/scala/shakey/server/internal/analyzer/R20Analyzer.scala
package shakey.server.internal.analyzer

import java.util

import org.apache.tapestry5.json.{JSONArray, JSONObject}
import shakey.server.internal.algorithm.StockAlgorithm
import shakey.server.services.StockAnalyzer

import scala.collection.mutable.ListBuffer

/**
 * 分析触碰到R20的股票
 */
class R20Analyzer extends StockAnalyzer {
  private val list = new ListBuffer[R20Stock]()

  def processStockDataInOneYear(symbol: String, data: JSONArray) {
    val pos: Int = data.length - 1
    val r20 = 0.until(20).
      map(x => data.getJSONObject(pos - x).
      getDouble("c")).sum / 20

    val current: JSONObject = data.getJSONObject(pos)
    val o = current.getDouble("o")
    val c = current.getDouble("c")
    val isReach = (Math.abs(o - r20) + Math.abs(r20 - c)) == Math.abs(c - o)
    if (isReach) {
      val rate = StockAlgorithm.calStrongRate(data, 25)
      if (rate > 0.003)
        list += new R20Stock(symbol, rate)
    }
  }

  def getTemplatePath: String = {
    return "/r20.ftl"
  }

  def addDataToTemplateAfterAnalysis(model: util.HashMap[AnyRef, AnyRef]) {
    val arr = list.sorted.toArray
    model.put("stocks", arr)
  }

  class R20Stock(val symbol: String, val strongRate: Double) extends Comparable[R20Stock] {
    def compareTo(o: R20Stock): Int = {
      o.strongRate.compareTo(strongRate)
    }
  }

}

File: shakey-server/src/main/java/shakey/server/internal/analyzer/R20Analyzer.java
package shakey.server.internal.analyzer;

import org.apache.tapestry5.json.JSONArray;
import org.apache.tapestry5.json.JSONObject;
import shakey.server.internal.stat.StockAlgorithm;
import shakey.server.services.StockAnalyzer;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;

/**
 * 计算那些股票触碰了20日均线
 *
 * @author jcai
 */
public class R20Analyzer implements StockAnalyzer {
    public void processStockDataInOneYear(String symbol, JSONArray data) {
        double r20 = 0.0;
        int pos = data.length() - 1;
        for (int i = 0; i < 20; i++) {
            r20 += data.getJSONObject(pos - i).getDouble("c");
        }
        r20 /= 20;

        JSONObject current = data.getJSONObject(pos);
        double o = current.getDouble("o");
        double c = current.getDouble("c");
        boolean isReach = (Math.abs(o - r20) + Math.abs(r20 - c)) == Math.abs(c - o);
        if (isReach) {
            list.add(new R20Stock(symbol, StockAlgorithm.calStrongRate(data, 25)));
        }
    }

    public String getTemplatePath() {
        return "/r20.ftl";
    }

    public void addDataToTemplateAfterAnalysis(HashMap<Object, Object> model) {
        R20Stock[] arr = list.toArray(new R20Stock[list.size()]);
        Arrays.sort(arr);
        model.put("stocks", arr);
    }

    private List<R20Stock> list = new ArrayList<R20Stock>();

    public static class R20Stock implements Comparable<R20Stock> {
        private String symbol;
        private double strongRate;

        public R20Stock(String symbol, double strongRate) {
            this.symbol = symbol;
            this.strongRate = strongRate;
        }

        public int compareTo(R20Stock o) {
            return Double.compare(o.strongRate, strongRate);
        }

        public String getSymbol() {
            return symbol;
        }

        public void setSymbol(String symbol) {
            this.symbol = symbol;
        }

        public double getStrongRate() {
            return strongRate;
        }

        public void setStrongRate(double strongRate) {
            this.strongRate = strongRate;
        }
    }
}
